### PR TITLE
Improve accuracy of System 573's digital I/O audio emulation

### PR DESCRIPTION
**Pull request by 987123879113**
This PR is an attempt to improve accuracy of the System 573's digital I/O audio emulation. 

I noticed only 2 other devices in all of MAME even use STREAM_SYNCHRONOUS and I realize there's performance hit involved with its usage so I will offer a justification for its usage here. If there are other options besides STREAM_SYNCHRONOUS that would allow for accurately tracking playback at the sample level without the performance hit then I'd love suggestions.

The biggest issue currently with the k573dio/fpga emulation is audio buffering. The current sample counter implementation relies on the buffered sample counts: `u32 get_frame_count() const { return total_frame_count - buffered_frame_count; }`.

With this implementation, the sample counter increments in fixed intervals whenever the buffer is filled by sound_stream_update. I tried looking around in the sound-related interfaces and could only really find functions that returned the start and end positions of the current buffer but not the exact position being played within that buffer.

I ran various tests on real hardware to confirm that it's possible to read the counter at a per-sample level. I also checked that it's directly related to the number of samples and not something else like a standard timer just to be absolutely sure. On real hardware the sample counter increases proportional to the sample rate of the MP3. For example, 48 kHz will increase faster than 44.1kHz, and a lower sample rate causes the sample counter to increase slower than the sample counter at 44.1 kHz.

One test I did that highlights the issue the clearest is a test that reads the sample counter register multiple times in succession with a vsync call in between. Using a MP3 at 44.1 kHz as an example here.
MAME results: The difference of sample counter updates in between vsyncs is always 882 with no variation.
Real hardware results: The difference of sample counter updates in between vsyncs is roughly around 736 with a variance of roughly 15.

Konami uses the sample counter heavily for key parts of gameplay (animations and song timing), and the difference between what MAME returns and what the game expects results in noticeably inaccurate timings. As such, I turned the mas3507d device into a synchronous audio device so that every sample can be counted and an accurate sample count can be returned. This change brings the emulated code much closer to real hardware.

### CODE REVIEW

**Comment by galibert** [comment]
Are you saying that mame has the vsync at 50Hz and your board at 59.92Hz ?

Also, a good remark by smf, did you try not using sync but just calling
stream->update() before reading the counter?

  OG.

**Comment by 987123879113** [comment]
>Are you saying that mame has the vsync at 50Hz and your board at 59.92Hz ?

I'm not comfortable stating that exact conclusion in those words. I regret not trying to approach this PR back when the research was fresher. I did all of this research a few months ago so exact details escape me, but I found the videos I recorded from real hardware. Hopefully those will help illustrate my points better.

Test 1. This begins when the words "Sound Scale Check 2" turn red. The value on the left is the sample counter register. The value on the right is the calculated difference since last update (NOT the diff register since I hadn't found that yet). This test is not using the vsync method but instead just letting the game run naturally. This is unfortunately the only time I recorded the results of both MAME and real hardware using the same code. You can see the difference clearly here.

Real hardware: https://streamable.com/mgg6zj
The counter jitters roughly around 0x2e0. 

MAME: https://streamable.com/5pa183
The value is either 0x372 or 0. I believe the 0 reads are when the game calls the register again before the MAME actually finished reading the entire buffered sample data and requests more samples. In that case, the sample counter would be the same between reads.

---

Test 2. The left counter is the FPGA's sample counter, the right counter is the MAS3507D's default read (frame) counter. Other things of interest here are the bottom left corner values. The one on the left corresponds to fpga_ctrl_r, the one on the right corresponds to mpeg_ctrl_r. Sorry for the lack of audio. It's the same scale test file as the other videos. I was trying to figure out how much the FPGA's sample counter ticked between the reported frame counter from MAS3507D's default read.

Real hardware: https://streamable.com/trlorr

---

Test 3.

1. Fast - Read the same register 4 times with minimal code in between (just only a few cycles wasted between each read, so not 100% instantly after but not a great delay between reads)
2. Loop delay - After each read, waste cycles by forcing the CPU to decrement a counter from 0x1000 down to 0 before moving on to the next read.
3. VSync delay - Call VSync() after every read.

This test could have been conducted better, but I will leave it in because it's all I have recorded.

Real hardware, polling the counter register: https://streamable.com/t1tzkl
Real hardware, polling mp3_counter_diff_r: https://streamable.com/dj9815

Interesting thing I had noted for this test: even with my changes to the registers which did roughly fix the "fast" and "vsync delay" tests, the loop delay test was actually off by a significant margin. I think the loop test in MAME was actually reporting something like ~0x20 instead of ~0x10 like real hardware, which could imply some other timing issues exist elsewhere (PSX emu-related code?) but I don't think that is related to this PR exactly and it's quite out of my league. 

--- 

I managed to short my digital I/O board after that when I was tracing connections around the board so I haven't been able to do any further hardware testing. 

>Also, a good remark by smf, did you try not using sync but just calling stream->update() before reading the counter?

That's a good one. I didn't realize that was an option. I assumed that I had to leave updating up to the backend. I'll give it a try.

**Comment by 987123879113** [comment]
@galibert I found my build of test 3 with polling the counter register so I recorded a video of MAME using this PR vs MAME without this PR.

Real hardware: https://streamable.com/t1tzkl
MAME, this PR: https://streamable.com/nmi98c
MAME, current  (0.227): https://streamable.com/a2xnxr

In the MAME this PR video I tried to highlight the loop test issue I mentioned at the end of the video, where you can clearly see that the loop test is resulting in a difference twice as great as real hardware (~0x20 instead of ~0x10) just for reference.

In the MAME 0.227 video you can see:
- Default read is not implemented so the one counter constantly returns 0xffff (does not affect gameplay though from what I can tell)
- In the bottom left corner, the register on the right is not implemented at all
- If you pause the video at any point of any of the tests, the value between the 4 different register reads never actually changes in the fast and loop tests, and doesn't change much between vsyncs in the vsync delay test. That's due to the buffering issue that I mentioned. The counter is only being updated every time new samples are buffered.

I also whipped up a quick Python tool to create a cheat file between the EXE diffs. This should let you reproduce the test locally. I've attached the cheat file as well as the original ASM patch file (meant to be compiled with the armips assembler).

Download: [ddrextrm-test.zip](https://github.com/mamedev/mame/files/5812120/ddrextrm-test.zip)



I hope this helps.

**Comment by 987123879113** [comment]
>Also, a good remark by smf, did you try not using sync but just calling stream->update() before reading the counter?

This turned out to work really well. After doing some refactoring I was able to ditch `STREAM_SYNCHRONOUS` and still keep the more accurate feel to the counters. Aside from the loop test having a difference of ~0x20 every update, everything else is roughly the same.

MAME, using forced stream updates: https://streamable.com/sw4yp5

Please let me know what you think about the current changes.

File: src/mame/machine/k573dio.cpp [hunk 1]
```diff
@@ -120,8 +133,15 @@ void k573dio_device::device_start()
 {
 	output_cb.resolve_safe();
 
-	ram = std::make_unique<uint16_t[]>(0x2000000/2);
-	save_pointer(NAME(ram), 0x2000000/2 );
+	const int ram_size = 0x2000000/2;
+	ram = std::make_unique<uint16_t[]>(ram_size);
```
Comment (cuavas) [inline src/mame/machine/k573dio.cpp:137]: Could you please change this to a `memory_share_creator` in the DIO device, and then use either another `memory_share_creator` or a `required_memory_share` or another `memory_share_creator` in the FPGA device to find it?

File: src/devices/sound/mas3507d.cpp [hunk 2]
```diff
@@ -61,6 +100,14 @@ void mas3507d_device::i2c_scl_w(bool line)
 		if(i2c_bus_state == STARTED) {
 			if(i2c_sdai)
 				i2c_bus_curval |= 1 << i2c_bus_curbit;
+
+			if (i2c_subdest == DATA_READ) {
```
[the remark this answers is not in the document]
Comment (987123879113) [reply src/devices/sound/mas3507d.cpp:104]: Sure, no problem. I think the other files probably have the same issues involving spacing and most likely bracket usage as well. There's some others I noticed and probably others that aren't shown in the diffs, so I'll give all of the files I touched in this PR a style pass to make sure they're consistent with the existing style.

File: src/mame/machine/k573dio.cpp [hunk 3]
```diff
@@ -131,8 +151,9 @@ void k573dio_device::device_reset()
 {
 	ram_adr = 0;
 	ram_read_adr = 0;
+	crypto_key1 = 0;
 
-	memset(output_data, 0, sizeof(output_data));
+	std::fill_n(output_data, sizeof(output_data), 0);
```
[the remark this answers is not in the document]
Comment (987123879113) [reply src/mame/machine/k573dio.cpp]: I didn't realize that. I'll be more mindful in the future. Thanks for the info.

File: src/devices/sound/mas3507d.h [hunk 4]
```diff
@@ -22,48 +28,73 @@ class mas3507d_device : public device_t, public device_sound_interface
...
-
-	std::array<u8, 0xe00> mp3data;
-	std::array<mp3d_sample_t, MINIMP3_MAX_SAMPLES_PER_FRAME> samples;
-	bool i2c_scli, i2c_sclo, i2c_sdai, i2c_sdao;
-	int i2c_bus_curbit;
-	uint8_t i2c_bus_curval;
-	int mp3_count, sample_count, current_rate;
-	u32 total_frame_count, buffered_frame_count;
+	enum { UNDEFINED, CONTROL, DATA_READ, DATA_WRITE, BAD } i2c_subdest;
+	enum { CMD_BAD, CMD_RUN, CMD_READ_CTRL, CMD_WRITE_REG, CMD_WRITE_MEM, CMD_READ_REG, CMD_READ_MEM } i2c_command;
```
[the remark this answers is not in the document]
Comment (987123879113) [reply src/devices/sound/mas3507d.h]: That's how it was before so I can't answer that.

I made the decision to split off the anonymous enum and the variable itself so in my latest commits. Previously save states for those variables weren't possible due to the anonymous enum, but now it works so I added them to the save states. 

I also stopped being lazy and add the complex types for minimp3 that I skipped when adding save states. I think this should make the mas3507d save state complete now. As part of this change, I decided to update the minimp3 library to latest master since there's about 2-3 years worth of updates. That also gets rid of some old hacks I previously put in the minimp3 library so updating/maintenance should be easier in the future.

File: src/devices/sound/mas3507d.h [hunk 5]
```diff
@@ -67,10 +67,15 @@ class mas3507d_device : public device_t, public device_sound_interface
...
-	enum { CMD_BAD, CMD_RUN, CMD_READ_CTRL, CMD_WRITE_REG, CMD_WRITE_MEM, CMD_READ_REG, CMD_READ_MEM } i2c_command;
+	enum { IDLE = 0, STARTED, NAK, ACK, ACK2 };
+	enum { UNKNOWN = 0, VALIDATED, WRONG };
+	enum { UNDEFINED = 0, CONTROL, DATA_READ, DATA_WRITE, BAD };
+	enum { CMD_BAD = 0, CMD_RUN, CMD_READ_CTRL, CMD_WRITE_REG, CMD_WRITE_MEM, CMD_READ_REG, CMD_READ_MEM };
+
+	uint8_t i2c_bus_state;
+	uint8_t i2c_bus_address;
+	uint8_t i2c_subdest;
+	uint8_t i2c_command;
```
Comment (cuavas) [inline src/devices/sound/mas3507d.h:78]: For what it’s worth, you can make enumerated types that can be saved.  You declare it like:

```
enum i2c_bus_state_t : uint8_t { IDLE = 0, STARTED, NAK, ACK, ACK2 };
...
i2c_bus_state_t i2c_bus_state;
```

Then in your `.cpp` file you use a helper macro to declare the stuff allowing it to be saved:
```
ALLOW_SAVE_TYPE(mas3507d_device::i2c_bus_state_t)
```

(Remember to avoid usin names ending in `_t` at global scope because they’re reserved by POSIX.  They’re fine at class scope, though.)

It’s OK how it is now, just pointing out that you can keep the enum type information and still save things if you want to.

File: src/mame/machine/k573fpga.cpp [hunk 6]
```diff
@@ -13,6 +13,7 @@
 
 k573fpga_device::k573fpga_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
 	device_t(mconfig, KONAMI_573_DIGITAL_FPGA, tag, owner, clock),
+	ram(*this, ":k573dio:ram"),
```
Comment (cuavas) [inline src/mame/machine/k573fpga.cpp:16]: Once again, this makes assumptions about the system hosting it, which is something we try to avoid.  The preferred way to do this kind of connection would be…

Construct it here as:
```
ram(*this, finder_base::DUMMY_TAG),
```

In the header provide a configuration method for setting the tag:
```
template <typename T> void set_ram(T &&tag) { ram.set_tag(std::forward<T>(tag)); }
```

Then when you instantiate it in the DIO device, you should be able to do this:
```
KONAMI_573_DIGITAL_FPGA(config, k573fpga);
k573fpga->set_ram(ram);
```

That way, the literal tag for the RAM is only in one place, you should get a validity check error on start if the RAM isn’t found, and it’s not relying on where any of the devices are in relation to each other.

File: src/devices/sound/mas3507d.cpp [hunk 7]
```diff
@@ -78,6 +78,21 @@ void mas3507d_device::device_start()
...
+	save_item(STRUCT_MEMBER(mp3_dec, free_format_bytes));
+	save_item(STRUCT_MEMBER(mp3_dec, header));
+	save_item(STRUCT_MEMBER(mp3_dec, reserv_buf));
+
+	save_item(STRUCT_MEMBER(mp3_info, frame_bytes));
+	save_item(STRUCT_MEMBER(mp3_info, frame_offset));
+	save_item(STRUCT_MEMBER(mp3_info, channels));
+	save_item(STRUCT_MEMBER(mp3_info, hz));
+	save_item(STRUCT_MEMBER(mp3_info, layer));
+	save_item(STRUCT_MEMBER(mp3_info, bitrate_kbps));
```
Comment (cuavas) [inline src/devices/sound/mas3507d.cpp:95]: You don’t actually need to use `STRUCT_MEMBER` for these, `NAME(mp3_dec.mdct_overlap)` etc. will work fine.  `STRUCT_MEMBER` provides glue for saving a member of a struct within an array as a single safe state item.  It's for stuff like this:
```
struct foo_t {
    u8 x, y;
};
foo_t foo[16];
...
save_item(STRUCT_MEMBER(foo, x));
save_item(STRUCT_MEMBER(foo, y));
```

It makes members of the structure in an array look like arrays to the save state system and the debugger.  Since `mp3_dec` and `mp3_info` aren’t inside arrays, you don’t need the extra magic.

**Comment by cuavas** [comment]
Since the commits are fairly self-contained changes, I commented on them individually.  The only one that’s I’d like to see changed before approving this is the one for how the RAM finder in the FPGA device is configured.  The other ones are more general commentary, and not required changes.

File: src/mame/machine/k573fpga.cpp (hunk 6 above)
Comment (987123879113) [reply src/mame/machine/k573fpga.cpp:16]: I understand what you were asking for now. I've implemented the code but there's an error and I'm not comfortable making the (seemingly?) required change without approval.

`k573fpga->set_ram(ram);` won't compile since `set_ram` wants a tag, not the actual `memory_share_creator`. My assumption is that I would need to make this code `k573fpga->set_ram(ram.finder_tag());` to actually get the tag from the `memory_share_creator`. However, it doesn't seem like it's actually possible to get the tag from `memory_share_creator` due to access level issues. 

`memory_share_creator` is declared as `class memory_share_creator : finder_base` which is making everything inside `finder_base` private. Changing it to `class memory_share_creator : public finder_base` allows me to use `finder_tag()` from `finder_base` but I don't have a deep enough understanding of the codebase to really understand if there are any unwanted implications involved with exposing `finder_base` functions here.

Or I could just use a const string for the tag in k573dio.

Please let me know what you think about the proposed change to `memory_share_creator`.

File: src/devices/sound/mas3507d.h (hunk 5 above)
Comment (987123879113) [reply src/devices/sound/mas3507d.h:78]: This actually is how I wanted to do it but ran into errors about not being able to save the items in save states. I'll make the changes, as well as the ones to remove STRUCT_MEMBER.

File: src/mame/machine/k573fpga.cpp (hunk 6 above)
Comment (cuavas) [reply src/mame/machine/k573fpga.cpp:16]: It doesn’t require a literal tag, it can take on the target of another object finder.  Notice the overload `finder_base::set_tag(finder_base const &)`: https://github.com/mamedev/mame/blob/mame0227/src/emu/devfind.h#L304

We use this pattern pretty heavily to avoid repeating literal tags.

Private inheritance for `memory_share_creator` and `memory_bank_creator` is indeed an issue.  I think this is just an oversight – it should be fine to just change them both to use public inheritance.  Unfortunately it will trigger a recompilation of everything, since it’s part of the precompiled header.

If that change offends @galibert, I’ll take the heat.

Comment (987123879113) [reply src/mame/machine/k573fpga.cpp:16]: You're right, I just tried doing it the way you wrote in the code snippet and it's compiling just fine. I guess the private inheritance was what was causing it to not compile in the first place and I misinterpreted the error message. 

I'll make the change to `memory_share_creator` as well as `memory_bank_creator` then.

Thanks.

**Review by cuavas — APPROVED** [review]

**Comment by cuavas** [comment]
Some people seem pretty happy with this: https://www.reddit.com/r/MAME/comments/kyracw/checking_the_new_accuracy_improvements_of_system/

**Comment by 987123879113** [comment]
Awesome, thanks for the link. Glad people are enjoying it. This PR, along with #7659, should make System 573 fans (*especially* System 573-era DDR fans) really happy when the next version of MAME is released. Probably most casual players in the west didn't actually use the memory card reader because the requirements were rather high for your typical casual player (specific arcade cabinet configuration required, and must own the Japanese PS1 copy of the game to even make the required save data to use in the arcade) but score saving is pretty huge and makes playing for scores a lot more enjoyable. Plus this PR in particular fixes timing issues and latency a significant amount. I will probably revisit this in the future once I can eventually get more hardware testing done since there are still improvements that can be made.

In the mean time, I'm working on cleaning up and preparing some other PRs that should make a small chunk of currently unplayable Bemani music games fully playable in MAME. So Bemani/music game fans are in for more treats this year. :D
